crates/sui-core/src/authority/authority_store.rs

```diff
@@ -144,6 +144,36 @@ impl AuthorityStore {
         Ok(store)
     }
 
+    /// This function is called at the very end of the epoch.
+    /// This step is required before updating new epoch in the db and calling reopen_epoch_db.
+    pub(crate) async fn revert_uncommitted_epoch_transactions(&self) -> SuiResult {
+        let epoch_store = self.epoch_store.load();
+        {
+            let state = epoch_store.get_reconfig_state_write_lock_guard();
+            if state.should_accept_user_certs() {
+                // Need to change this so that consensus adapter do not accept certificates from user.
```

Comment:
Here it seems could be a potential race where a cert slipped through and is being executed at the same time, but not yet added to  pending_consensus_certificates

Reply:
So ConsensusAdapter(or more specifically AuthorityServer::handle_certificate) is the only place that can accept certificate that might need to get reverted.
Other places that can take certificates, both state sync and narwhal only accept certificates that are non-revertable.

I think I agree that `handle_certificate` does need a bit of change to make it bullet proof, more specifically we need to move state locking out of ConsenusAdapter to the AuthorityServer and check the lock state before we execute.

Is that what you mean by a potential race or something else come to mind? I can fix this, either on this PR on separately (which I would prefer to keep PRs small)

Reply:
Yes!